### PR TITLE
Move files for release (issue #193)

### PR DESCRIPTION
**Pull request by yolaj-nhs**
Add a function into attack_report_formatter.py which will move all of the files generated by the release process into two folders:

- training_artefacts (or similar name) - files generated in testing process (not released)
- release_files - files to be released, which will contain four files:
   - the JSON of the attack output
   - the JSON of the target
   - the model to be released
   - the final .txt recommendation

### CODE REVIEW

File: aisdc/attacks/attack_report_formatter.py [hunk 1]
```diff
@@ -5,12 +5,38 @@
...
+    Function that will move any files created throughout the release process and
+    sort them into appropriate folders
+    """
+
+    if not os.path.exists(release_dir):
+        os.makedirs(release_dir)
+
+    for filename in move_into_artefacts:
+        if os.path.exists(filename):
+            dest = artefacts_dir + "/" + os.path.basename(filename).split("/")[-1]
```
Comment (simonrnss) [inline aisdc/attacks/attack_report_formatter.py:31]: safer to use `os.sep` instead of the explicit use of "/"?

Comment (simonrnss) [reply aisdc/attacks/attack_report_formatter.py:31]: same in other places below

File: aisdc/attacks/attack_report_formatter.py [hunk 2]
```diff
@@ -486,6 +514,10 @@ def _process_target_json(self):
                     )
                     output_string = output_string + "\n"
 
+        if "model_path" in json_report.keys():
+            filepath = os.path.split(os.path.abspath(self.target_json_filename))[0]
+            self.model_name_from_target = filepath + "\\" + json_report["model_path"]
```
Comment (simonrnss) [inline aisdc/attacks/attack_report_formatter.py:519]: `"\\"` now?

File: example_notebooks/user_stories/user_story_1/user_story_1_tre.py [hunk 3]
```diff
@@ -24,9 +24,9 @@ def generate_report(directory, attack_results, target, outfile):
         directory + attack_results, target_filename=directory + target
     )
 
-    t.export_to_file(output_filename=outfile)
+    t.export_to_file(output_filename=directory + outfile, move_files=True)
```
Comment (simonrnss) [inline example_notebooks/user_stories/user_story_1/user_story_1_tre.py:27]: does this assume that the separator is in the directory variable? I'd prefer `os.path.join` for stuff like this

**Review by simonrnss — CHANGES_REQUESTED** [review]
Generally fine, but I'm a little uncomfortable about the use of `+` for forming paths and explicit encoding of path seperators. `os.path.join` and `os.sep` would probably be safer? But, happy to go with what's been done if it matches the format elsewhere in the codebase

**Comment by yolaj-nhs** [comment]
Thanks @simonrnss for your comments - you're absolutely right. Changes made

**Comment by yolaj-nhs** [comment]
Converting to draft while I fix code coverage

File: aisdc/attacks/attack_report_formatter.py [hunk 4]
```diff
@@ -544,7 +584,14 @@ def process_attack_target_json(
 
         self.text_out.append(bucket_text)
 
-    def export_to_file(self, output_filename: str = "summary.txt"):
+    def export_to_file(  # pylint: disable=too-many-arguments
+        self,
+        output_filename: str = "summary.txt",
+        move_files=False,
+        model_filename=None,
+        release_dir="release_files/",
```
Comment (simonrnss) [inline aisdc/attacks/attack_report_formatter.py:592]: do we need the trailing /?

Comment (simonrnss) [reply aisdc/attacks/attack_report_formatter.py:592]: ditto below

File: tests/test_attack_report_formatter.py [hunk 5]
```diff
@@ -201,6 +201,69 @@ def test_whitespace_in_filenames(self):
...
+
+        with open(dummy_model, "w", encoding="utf-8") as f:
+            f.write("dummy_model")
+
+        g = GenerateTextReport()
+        g.process_attack_target_json(filename)
+        g.export_to_file(output_filename, move_files=True, release_dir="release_dir/")
+
+        # Check when no model name has been provided
+        assert os.path.exists("release_dir/" + output_filename) is True
```
Comment (simonrnss) [inline tests/test_attack_report_formatter.py:223]: maybe change to `os.path.exists(os.path.join(...`

File: tests/test_attack_report_formatter.py [hunk 6]
```diff
@@ -201,6 +201,69 @@ def test_whitespace_in_filenames(self):
...
+
+        with open(filename, "w", encoding="utf-8") as f:
+            json.dump(json_formatted, f)
+
+        with open(dummy_model, "w", encoding="utf-8") as f:
+            f.write("dummy_model")
+
+        g = GenerateTextReport()
+        g.process_attack_target_json(filename)
+        g.export_to_file(output_filename, move_files=True, release_dir="release_dir/")
```
Comment (simonrnss) [inline tests/test_attack_report_formatter.py:220]: remove `/`

File: tests/test_attack_report_formatter.py [hunk 7]
```diff
@@ -201,6 +201,69 @@ def test_whitespace_in_filenames(self):
...
+
+        g.export_to_file(
+            output_filename,
+            move_files=True,
+            release_dir="release_dir/",
+            model_filename=dummy_model,
+        )
+
+        # Check model file has been copied (NOT moved)
+        assert os.path.exists("release_dir/" + dummy_model) is True
```
Comment (simonrnss) [inline tests/test_attack_report_formatter.py:233]: ditto

File: tests/test_attack_report_formatter.py [hunk 8]
```diff
@@ -201,6 +201,69 @@ def test_whitespace_in_filenames(self):
...
+            pass
+
+        assert os.path.exists(png_file) is True
+
+        g = GenerateTextReport()
+        g.process_attack_target_json(filename)
+        g.export_to_file(
+            output_filename,
+            move_files=True,
+            release_dir="release_dir/",
```
Comment (simonrnss) [inline tests/test_attack_report_formatter.py:251]: ditto

File: tests/test_attack_report_formatter.py [hunk 9]
```diff
@@ -201,6 +201,69 @@ def test_whitespace_in_filenames(self):
...
+            artefacts_dir="training_artefacts/",
+        )
+
+        assert os.path.exists(png_file) is False
+        assert os.path.exists("training_artefacts/" + png_file) is True
+
+        clean_up(filename)
+        clean_up(output_filename)
+        clean_up("release_dir/" + dummy_model)
```
Comment (simonrnss) [inline tests/test_attack_report_formatter.py:260]: `os.path.join`

**Review by simonrnss — APPROVED** [review]
Just some more bits where we're explicitly adding dir seps and not using os.path.join

**Comment by yolaj-nhs** [comment]
@shahzadmumtaz22 I'm merging this into development just so it's out before the end of sprint, I'm happy to make a new issue if you have any comments

**Pull request closed without merging**